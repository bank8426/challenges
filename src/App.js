import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as charityActions from './redux/actions/charityActions';
import * as paymentActions from './redux/actions/paymentActions';
import styled from 'styled-components';
import fetch from 'isomorphic-fetch';
import { bindActionCreators } from 'redux'

const Card = styled.div`
  margin: 10px;
  border: 1px solid #ccc;
`;

function mapStateToProps(state) {
  return {
    charities: state.charities,
    loading: state.apiCallsInProgress > 0,
    donate: state.donate,
    message: state.message,
    payments: state.payments,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: {
      loadCharities: bindActionCreators(charityActions.loadCharities, dispatch),
      loadPayments: bindActionCreators(paymentActions.loadPayments, dispatch),
    },
    dispatch,
  };
}

export default connect(mapStateToProps,mapDispatchToProps)(
  class App extends Component {
    constructor(props) {
      super();

      this.state = {
        selectedAmount: 10,
      };
    }

    componentDidMount() {
      const { charities, actions } = this.props;
      if (charities.length === 0) {
        actions.loadCharities().catch(error => {
          alert('Loading charities failed' + error);
        });
      }

      actions.loadPayments().catch(error => {
        alert('Loading payments failed' + error);
      })
    }

    render() {
      const self = this;
      const cards = this.props.charities.map(function(item, i) {
        const payments = [10, 20, 50, 100, 500].map((amount, j) => (
          <label key={j}>
            <input
              type="radio"
              name="payment"
              onClick={function() {
                self.setState({ selectedAmount: amount })
              }} /> {amount}
          </label>
        ));
        
        return (
          <Card key={i}>
            <p>{item.name}</p>
            <img src={process.env.IMAGES_PATH + item.image} />
            {payments}
            <button onClick={handlePay.call(self, item.id, self.state.selectedAmount, item.currency)}>Pay</button>
          </Card>
        );
      });

      const style = {
        color: 'red',
        margin: '1em 0',
        fontWeight: 'bold',
        fontSize: '16px',
        textAlign: 'center',
      };
      const donate = this.props.donate;
      const message = this.props.message;

      return (
        <div>
          <h1>Tamboon React</h1>
          {this.props.loading ? (
            <div>Loading</div>
          ) : (
            <>
              <p>All donations: {donate}</p>
              <p style={style}>{message}</p>
              {cards}
            </>
          )}
        </div>
      );
    }
  }
);

function handlePay(id, amount, currency) {
  const self = this;
  return function() {
    fetch('http://localhost:3001/payments', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: `{ "charitiesId": ${id}, "amount": ${amount}, "currency": "${currency}" }`,
    })
      .then(function(resp) { return resp.json(); })
      .then(function() {
        self.props.dispatch({
          type: 'UPDATE_TOTAL_DONATE',
          amount,
        });
        self.props.dispatch({
          type: 'UPDATE_MESSAGE',
          message: `Thanks for donate ${amount}!`,
        });

        setTimeout(function() {
          self.props.dispatch({
            type: 'UPDATE_MESSAGE',
            message: '',
          });
        }, 2000);
      });
  }
}
