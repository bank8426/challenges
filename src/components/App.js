import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as charityActions from '../redux/actions/charityActions';
import * as paymentActions from '../redux/actions/paymentActions';
import { bindActionCreators } from 'redux'
import Message from './common/Message';
import Donation from './common/Donation';
import Header from './common/Header';
import Spinner from './common/Spinner';
import CharityList from './Charity/CharityList';
import { GlobalStyles } from '../styles/global';
class App extends Component {
  componentDidMount() {
    const { charities, loadCharities , loadPayments } = this.props;
    if (charities.length === 0) {
      loadCharities().catch(error => {
        alert('Loading charities failed' + error);
      });
    }

    loadPayments().catch(error => {
      alert('Loading payments failed' + error);
    })
  }

  render() {
    const {donate,message,charities,isLoading} = this.props;
    return (
      <>
        <GlobalStyles />
        <Header/>
        {isLoading ? (
          <Spinner/>
        ) : (
          <>
            <Donation donate={donate} />
            <CharityList charities={charities}/>
            <Message message={message}/>
          </>
        )}
      </>
    );
  }
}

function mapStateToProps(state) {
  return {
    charities: state.charities,
    isLoading: state.apiCallsInProgress > 0,
    donate: state.donate,
    message: state.message,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    loadCharities: bindActionCreators(charityActions.loadCharities, dispatch),
    loadPayments: bindActionCreators(paymentActions.loadPayments, dispatch),
  };
}

export default connect(mapStateToProps,mapDispatchToProps)(App);