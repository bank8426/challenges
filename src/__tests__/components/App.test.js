import React from 'react';
import ReactDOM from 'react-dom';
import { App,mapStateToProps,mapDispatchToProps } from '../../components/App';
import { shallow,mount } from 'enzyme';
import StyledSpinner from '../../components/common/Spinner'
import renderer from 'react-test-renderer'
import {charities} from '../../../tools/mockData'
it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('it should display App', () => {
  const tree = renderer.create(<App />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should called componentDidMount', () => {
  const mockFn = jest.fn();
  const componentDidMountSpy = jest.spyOn(App.prototype, 'componentDidMount');
  window.alert = () => {mockFn()};
  shallow(<App/>);
  expect(App.prototype.componentDidMount).toHaveBeenCalledTimes(1);
  componentDidMountSpy.mockClear();
});

it('should alert when call loadCharities fail', async () => {
  const asyncMockFn = jest.fn().mockRejectedValue(new Error('Async error'));
  const mockFn2 = jest.fn();
  window.alert = mockFn2();
  shallow(<App loadCharities={ async () => await asyncMockFn() }/>);
  expect(mockFn2).toHaveBeenCalledTimes(1);
});

it('should alert when call loadPayments fail', async () => {
  const asyncMockFn = jest.fn().mockRejectedValue(new Error('Async error'));
  const mockFn2 = jest.fn();
  window.alert = mockFn2();
  shallow(<App loadPayments={ async () => await asyncMockFn() }/>);
  expect(mockFn2).toHaveBeenCalledTimes(1);
});

it('should show Spinner when it loading state', () => {
  const wrapper = mount(<App isLoading={true}/>);
  expect(wrapper.find(StyledSpinner).length).toEqual(1);
});

it('should not show Spinner when it load finish', () => {
  const wrapper = mount(<App isLoading={false}/>);
  expect(wrapper.find(StyledSpinner).length).toEqual(0);
});

it('should dispatch action when loadCharities() is called', () => {
  const dispatch = jest.fn();
  const props = mapDispatchToProps(dispatch); 
  props.loadCharities()
  expect(dispatch).toHaveBeenCalledTimes(1);
});

it('should dispatch action when loadPayments() is called', () => {
  const dispatch = jest.fn();
  const props = mapDispatchToProps(dispatch);
  props.loadPayments()
  expect(dispatch).toHaveBeenCalledTimes(1);
});

it('should mapStateToProps', () => {
  const state = jest.fn();
  const props = mapStateToProps(state);
  props.donate = 0
  expect(props.donate).toEqual(0);
});

it('should display charities when loading finish and have data', () => {
  const initCharities = charities;
  const tree = renderer.create(<App isLoading={false} charities={initCharities} />).toJSON()
  expect(tree).toMatchSnapshot()
});