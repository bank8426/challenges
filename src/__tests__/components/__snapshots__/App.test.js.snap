// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should display App 1`] = `
Array [
  <h1
    className="sc-bxivhb lpgNEt"
  >
    Tamboon React
  </h1>,
  <div
    className="sc-ifAKCX jpzCyY"
  />,
]
`;

exports[`should display charities when loading finish and have data 1`] = `
Array [
  <h1
    className="sc-bxivhb lpgNEt"
  >
    Tamboon React
  </h1>,
  <h3
    className="sc-htpNat jvEGAF"
  >
    All donations: 
    0
     THB
  </h3>,
  <div
    className="sc-chPdSV gVUME"
  >
    <div
      className="sc-jTzLTM dMGDxY"
    >
      <div
        className="sc-fjdhpX hTYgkW"
        src="undefinedbaan-kru-noi.jpg"
      />
      <div
        className="sc-kAzzGY fLMWcz"
      >
        <h3
          className="sc-jzJRlG cQAfVf"
        >
          Baan Kru Noi
        </h3>
        <button
          className="primary sc-cSHVUG elibVm"
          onClick={[Function]}
        >
          Donate
        </button>
      </div>
    </div>
    <div
      className="sc-jTzLTM cTAeom"
    >
      <div
        className="sc-fjdhpX jnEfle"
        src="undefinedhabitat-for-humanity-thailand.jpg"
      />
      <div
        className="sc-kAzzGY fLMWcz"
      >
        <h3
          className="sc-jzJRlG cQAfVf"
        >
          Habitat for Humanity Thailand
        </h3>
        <button
          className="primary sc-cSHVUG elibVm"
          onClick={[Function]}
        >
          Donate
        </button>
      </div>
    </div>
    <div
      className="sc-jTzLTM hFNVOF"
    >
      <div
        className="sc-fjdhpX dMVtYd"
        src="undefinedpaper-ranger.jpg"
      />
      <div
        className="sc-kAzzGY fLMWcz"
      >
        <h3
          className="sc-jzJRlG cQAfVf"
        >
          Paper Ranger
        </h3>
        <button
          className="primary sc-cSHVUG elibVm"
          onClick={[Function]}
        >
          Donate
        </button>
      </div>
    </div>
    <div
      className="sc-jTzLTM jRZwSD"
    >
      <div
        className="sc-fjdhpX gQQjjf"
        src="undefinedmakhampom-theater.jpg"
      />
      <div
        className="sc-kAzzGY fLMWcz"
      >
        <h3
          className="sc-jzJRlG cQAfVf"
        >
          Makhampom Theater
        </h3>
        <button
          className="primary sc-cSHVUG elibVm"
          onClick={[Function]}
        >
          Donate
        </button>
      </div>
    </div>
    <div
      className="sc-jTzLTM hVnDW"
    >
      <div
        className="sc-fjdhpX fHcNRd"
        src="undefinedthailand-association-of-the-blind.jpg"
      />
      <div
        className="sc-kAzzGY fLMWcz"
      >
        <h3
          className="sc-jzJRlG cQAfVf"
        >
          Thailand Association of the Blind
        </h3>
        <button
          className="primary sc-cSHVUG elibVm"
          onClick={[Function]}
        >
          Donate
        </button>
      </div>
    </div>
  </div>,
  <div
    className="sc-bdVaJa cTeWxa"
  />,
]
`;
